src/libraries/System.Private.Xml/src/System/Xml/Serialization/Types.cs

```diff
@@ -1075,9 +1081,25 @@ private static void GetSettableMembers(StructMapping mapping, List<MemberMapping
                     {
                         throw new InvalidOperationException(SR.Format(SR.XmlReadOnlyPropertyError, propertyInfo.DeclaringType, propertyInfo.Name));
                     }
-                    list.Add(memberMapping);
+
+                    // Replace hidden base members with derived members, or else add the new member.
+                    if (!TryReplaceHiddenMapping(list, memberMapping))
+                        list.Add(memberMapping);
+                }
+            }
+        }
+
+        private static bool TryReplaceHiddenMapping(List<MemberMapping> list, MemberMapping memberMapping)
+        {
+            for (int i = 0; i < list.Count; i++)
+            {
+                if (memberMapping.Hides(list[i]))
+                {
```

Comment:
Does this code run on every serialization/deserialization or is it initialization code only? If it's on every serialization/deserialization, has the performance impact of this been measured? If you have a class with a lot of members, this is O(n^2) so has the potential to be a problem.

Reply:
For the ILGen serializer, this only runs once on initialization. Currently in the reflection-based implementation, this xml-mapping process happens for each call to serialize/deserialize. Which is one of the first things I'd look at when getting around to general reflection-serializer improvements, because it seems like it really only needs to happen once. But I did not address that here.

Reply:
FWIW, this is the xml-mapping code that I wouldn't want to bring back to previous versions.

Reply:
I just re-read my comment here and think I can state this a little better. It is the case for both serializers that when they are building their tree of code/IL/reflection-calls/whatever, they enter this code path to get member details for the type they are working with. In the case of the ILGen serializer it uses the results of this to generate IL that gets persisted in a TempAssembly and reused on each serialization, so it never needs to get this information again. In the case of the reflection-based serializer though, the reflection-call tree is not preserved. So this does get called multiple times. But this is the crux of the major perf improvements we want to make with the reflection serializer. When those updates come, this should only be called once.